### DISCUSSION TITLE
New standard for Pump It Up simfiles and data

### DISCUSSION THREAD

**Discussion by SheepyChris** [TeamRizu/OutFox#267] [03 - Feature Requests]
Creating this discussion after some talking over at Discord. Most PIU chart nowadays is created following a very loose standard introduced in StepF2:
- Difficulty is set to Edit, however sometimes they can be set to Easy or Medium;
- Description contains all information tags that will tell the theme what chart label to utilize (UCS, Mission, Train etc);
- Chart name contains any additional information that can be displayed if available, mostly utilized with Mission and Hidden charts.

The goal of this discussion is to ultimately modernize how charts will be made for the gamemode, bringing in improvements and more consistency to the library of content converted from the official games or made by users. There are currently a few ideas on what to do:
- Create a difficulty that is neutral and not considered as an "addition" to the simfile as how Edit works;
- Create a tag that will be reserved for labels and can be processed by the engine via enums to avoid string searching from themes and helping out with performance and consistency overall;
- Return to use the description tag as how it was originally meant to be used. 

These are just rapid thoughts on the standard, if you have any better ideas or solutions please post them down below!

**Comment by ListenerJubatus** [comment]
I think that any new step file format should _not_ lock difficulties to the DDR paradigm of only having five difficulties then edits. This goes for any game type...

**Comment by SHRMP0** [comment]
Posting this here too in case people not active on the Discord server join the discussion, since it may be useful for them:

Over the last months we made an effort to document how PIU simfiles/features work and keep track of their current state on OutFox and other forks. You can see our findings so far over [here](https://docs.google.com/document/d/1hMO1Vtw2iXgRdLoJJlZnhQEafnmKCjm1BuZ1duwZ_EQ/) and [here](https://docs.google.com/spreadsheets/d/1Q_--dCT4SgHL7CFSl6q8urKVTShdo-N6Gxkb3Trqb5s/). It's not 100% complete or up to date but hopefully can still serve as reference.

**Reply by djgrs** [reply]
Here some issues to solve (or not, at this moment) and adding new features:

- Holds problem (when you take a note with a hold, window time in the StepF2/P1/Prime as an example is HUGE, around 200 ms). In the 4.9.7 version of SM 5.3 apparently it's solved, although it's somewhat so hard you can't even wink or you'll get misses (some of people can't note the difference playing by hand obviously).

- When you play with Very Hard Judgement or above in the StepF2/P1, in a normal note attached with a long note you'll get a miss (i believe it's because timing window conflict or something like that...).

- Editor with mouse interaction (like ArrowVortex or DDReam). It will help a lot to make charts with effects!

- In Edit Mode, a feature you can extract your chart in image format (jpg, png or pdf). A lot of people study these charts printing into a paper.

- AV (MMod) issues concerning the display BPM (because of that, MMod can't work very well).

**Reply by snil4** [reply]
@SHRMP0 stepprime showed a working random notes placement based on festival of ghosts 2 mission [here](https://www.youtube.com/watch?v=2FKjbPXA28o), I didn't check if it's inside the public build yet but it seems to be working.
@djgrs most of your points are in the works including editor with mouse planned for alpha 5 and you can use AMod that takes the most common speed in the song (instead of the highest) and sets that as the display bpm (not perfect solution yet but should work better for pump).

**Reply by SHRMP0** [reply]
@snil4 thanks for the head's up, just updated the docs.

@djgrs some of the issues you described are currently only present on the Infinitesimal theme (most of the PIU-focused stuff is not yet implemented elsewhere), please make sure you test them again using SoundWaves (the default theme). Also use the latest 4.9.7GG build to prevent reporting already fixed behavior.
